File: Packages/RideSharingFeature/Sources/RideSharingFeature/DesignSystem/LiiveSwitch.swift
import SwiftUI

public struct LiiveSwitch: View {
    @Binding var isOn: Bool
    var disabled: Bool

    public init(isOn: Binding<Bool>, disabled: Bool = false) {
        self._isOn = isOn
        self.disabled = disabled
    }

    public var body: some View {
        Button {
            guard !disabled else { return }
            withAnimation(.easeOut(duration: LiiveMotion.base)) {
                isOn.toggle()
            }
        } label: {
            ZStack(alignment: .leading) {
                Capsule()
                    .fill(isOn ? LiiveColor.success : LiiveColor.fill)
                    .frame(width: LiiveSwitchLayout.trackWidth, height: LiiveSwitchLayout.trackHeight)
                Circle()
                    .fill(LiiveSwitchLayout.thumbColor)
                    .frame(width: LiiveSwitchLayout.thumbSize, height: LiiveSwitchLayout.thumbSize)
                    .shadow(
                        color: LiiveSwitchLayout.thumbShadowColor,
                        radius: LiiveSwitchLayout.thumbShadowRadius,
                        x: LiiveSwitchLayout.thumbShadowX,
                        y: LiiveSwitchLayout.thumbShadowY
                    )
                    .offset(x: isOn ? LiiveSwitchLayout.thumbTravel : LiiveSwitchLayout.thumbRestOffset)
                    .padding(LiiveSwitchLayout.trackPadding)
            }
            .opacity(disabled ? LiiveSwitchLayout.disabledOpacity : LiiveSwitchLayout.enabledOpacity)
        }
        .buttonStyle(.plain)
        .disabled(disabled)
        .accessibilityLabel(Text("Switch"))
        .accessibilityValue(Text(isOn ? "On" : "Off"))
        .accessibilityAddTraits(isOn ? .isSelected : [])
    }
}

private enum LiiveSwitchLayout {
    static let trackWidth = LiiveControl.lg + LiiveSpacing.xs2 / 2
    static let trackHeight = LiiveSpacing.xxxl - LiiveSpacing.xs2 / 2
    static let trackPadding = LiiveSpacing.xs2
    static let thumbSize = trackHeight - trackPadding - trackPadding
    static let thumbRestOffset = CGFloat.zero
    static let thumbTravel = trackWidth - thumbSize - trackPadding - trackPadding
    static let thumbShadowRadius = LiiveSpacing.xs
    static let thumbShadowX = CGFloat.zero
    static let thumbShadowY = LiiveSpacing.xs2
    static let thumbShadowColor = Color.black.opacity(0.25)
    static let thumbColor = Color.white
    static let disabledOpacity = 0.5
    static let enabledOpacity = 1.0
}

File: android/app/src/main/java/com/liive/ride/designsystem/LiiveSwitch.kt
package com.liive.ride.designsystem

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color

@Composable
fun LiiveSwitch(
    checked: Boolean,
    disabled: Boolean = false,
    onCheckedChange: (Boolean) -> Unit
) {
    val c = LiiveTheme.colors
    val trackColor by animateColorAsState(
        targetValue = if (checked) c.success else c.fill,
        animationSpec = tween(durationMillis = LiiveMotion.baseMs, easing = LiiveMotion.easeOut),
        label = "LiiveSwitchTrackColor"
    )
    val thumbOffset by animateDpAsState(
        targetValue = if (checked) LiiveSwitchLayout.ThumbTravel else LiiveSwitchLayout.ThumbRestOffset,
        animationSpec = tween(durationMillis = LiiveMotion.baseMs, easing = LiiveMotion.easeOut),
        label = "LiiveSwitchThumbOffset"
    )

    Box(
        Modifier
            .size(width = LiiveSwitchLayout.TrackWidth, height = LiiveSwitchLayout.TrackHeight)
            .alpha(if (disabled) LiiveSwitchLayout.DisabledOpacity else LiiveSwitchLayout.EnabledOpacity)
            .clip(LiiveRadius.full)
            .background(trackColor)
            .clickable(
                enabled = !disabled,
                interactionSource = remember { MutableInteractionSource() },
                indication = null
            ) { onCheckedChange(!checked) }
            .padding(LiiveSwitchLayout.TrackPadding),
        contentAlignment = Alignment.CenterStart
    ) {
        Box(
            Modifier
                .offset(x = thumbOffset)
                .size(LiiveSwitchLayout.ThumbSize)
                .shadow(LiiveSwitchLayout.ThumbShadowRadius, CircleShape)
                .clip(CircleShape)
                .background(LiiveSwitchLayout.ThumbColor)
        )
    }
}

private object LiiveSwitchLayout {
    val TrackWidth = LiiveControl.lg + LiiveSpacing.xs2 / 2
    val TrackHeight = LiiveSpacing.xxxl - LiiveSpacing.xs2 / 2
    val TrackPadding = LiiveSpacing.xs2
    val ThumbSize = TrackHeight - TrackPadding - TrackPadding
    val ThumbRestOffset = LiiveSpacing.xs2 - LiiveSpacing.xs2
    val ThumbTravel = TrackWidth - ThumbSize - TrackPadding - TrackPadding
    val ThumbShadowRadius = LiiveSpacing.xs
    val ThumbColor = Color.White
    const val DisabledOpacity = 0.5f
    const val EnabledOpacity = 1f
}
